### PR TITLE
evalengine: Proper support for bit literals

### PR DESCRIPTION
**Pull request by vmg**
## Description

We weren't handling bit literals properly, neither in the `sqlparser`, nor the `evalengine`.

The first problem we had was in the parser: the bind var extractor was normalizing all Bit literals as hex, which is not correct. There's an actual difference between hex and bits in MySQL: in a numerical context, hex is handled as an **unsigned** integer but bits are handled as a **signed** integer. So the parser must maintain the difference between them.

I normalized the parser to generate `BitNum` literals for all bit literals. Unlike for hex, where `HexVal` (`x'aabbccdd'`) and `HexNum` (`0xaabbccdd`) are handled differently during evaluation, both kinds of bit literals `0b0101` and `b'0101'` are handled identically, so we don't need to expose both literal types. We now normalize everything to the earlier "numeric" form with the `0b` prefix because it's easier to handle.

With these changes to the parser in place, we can update the `evalengine` to take into account the new types during evaluation.

cc @dbussink @systay 

## Related Issue(s)

- https://github.com/vitessio/vitess/pull/14337

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on the CI
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/mysql/collations/local.go [hunk 1]
```diff
@@ -49,9 +49,13 @@ func Default() ID {
...
 
 func DefaultCollationForType(t sqltypes.Type) ID {
+	return CollationForType(t, Default())
+}
+
+func CollationForType(t sqltypes.Type, default_ ID) ID {
 	switch {
 	case sqltypes.IsText(t):
-		return Default()
+		return default_
```
Comment (dbussink) [inline go/mysql/collations/local.go:58]: can the `_` be dropped?

File: go/vt/sqlparser/ast_format.go [hunk 2]
```diff
@@ -1314,12 +1314,10 @@ func (node *Literal) Format(buf *TrackedBuffer) {
 	switch node.Type {
 	case StrVal:
 		sqltypes.MakeTrusted(sqltypes.VarBinary, node.Bytes()).EncodeSQL(buf)
-	case IntVal, FloatVal, DecimalVal, HexNum:
+	case IntVal, FloatVal, DecimalVal, HexNum, BitNum:
```
Comment (dbussink) [inline go/vt/sqlparser/ast_format.go:1317]: I wonder, should we also refactor `HexVal` away? Or is there a legitimate reason we need to distinguish `HexNum` and `HexVal`? And if so, wouldn't that also apply to the `Bit` literals?

File: go/vt/vtgate/evalengine/api_type_aggregation.go [hunk 3]
```diff
@@ -57,6 +57,21 @@ func AggregateTypes(types []sqltypes.Type) sqltypes.Type {
...
+	var t sqltypes.Type
+	var f typeFlag
+	switch e := e.(type) {
+	case nil:
+		t = sqltypes.Null
+	case *evalBytes:
+		t = sqltypes.Type(e.tt)
+		f = e.flag
+	default:
+		t = e.SQLType()
```
Comment (dbussink) [inline go/vt/vtgate/evalengine/api_type_aggregation.go:70]: Do we not care about flags in this case?

File: go/vt/vtgate/evalengine/fn_time.go [hunk 4]
```diff
@@ -1375,7 +1375,7 @@ func dateTimeUnixTimestamp(env *ExpressionEnv, date eval) evalNumeric {
 			case *evalDecimal:
 				prec = d.length
 			case *evalBytes:
-				if d.isHexLiteral {
+				if d.isHexOrBitLiteral() {
 					return newEvalInt64(0)
```
Comment (dbussink) [inline go/vt/vtgate/evalengine/fn_time.go:1379]: Should this return a `uint64` in case of a hex literal?

**Review by systay — APPROVED** [review]
lgtm

File: go/vt/vtgate/evalengine/fn_time.go (hunk 4 above)
Comment (vmg) [reply go/vt/vtgate/evalengine/fn_time.go:1379]: Fixed

**Review by dbussink — APPROVED** [review]

**Pull request merged**